pylint/checkers/classes/special_methods_checker.py

```diff
@@ -308,55 +319,67 @@ def _is_iterator(node):
                     pass
         return False
 
-    def _check_iter(self, node, inferred):
+    def _check_iter(self, node: nodes.FunctionDef, inferred: Any) -> None:
         if not self._is_iterator(inferred):
             self.add_message("non-iterator-returned", node=node)
 
-    def _check_len(self, node, inferred):
+    def _check_len(self, node: nodes.FunctionDef, inferred: nodes.Const) -> None:
```

Comment:
Just picking this one as it is the easiest to illustrate, I don't really understand where these typings are coming from? Don't they all come from `_protocol_map` and called with `Callable[[nodes.FunctionDef, Any], None]`?

The `Any` being the result of `infer_call_result`? Why can this only be `Const`?

Reply:
For:
```
    def __len__(self):  # [invalid-length-returned]
        return lambda: 3
```
It's not a const but a ``<class 'astroid.nodes.scoped_nodes.scoped_nodes.Lambda>'. pyannotate must have a bug.

Reply:
I think this is not necessarily a bug with `pyannotate` but just shows that our test case doesn't cover all edge cases. Like I said, perhaps revert the changes here and keep this PR focused on those that are definetly correct?

Reply:
> our test case doesn't cover all edge cases.

Although true in the general case, I added an assert in ``def _check_len``, the example come from ``invalid_length_returned.py`` that is supposed to be launched during functional tests so either pyannotate has a bug or the way we launch tests is not understood.